aten/src/ATen/native/Linear.cpp

```diff
@@ -136,241 +136,319 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
-          if (num_ell_idxes == -1) {      // if we have not seen an ellipsis before, keep track of indices and size
-            first_ell_idx = num_total_idxes;
-            num_ell_idxes = candidate_num_ell_idxes;
-            num_total_idxes += num_ell_idxes;
-          }
-          else {                          // we have seen an ellipsis before, so we check compatibility
-            TORCH_CHECK(candidate_num_ell_idxes == num_ell_idxes,
-                     "ellipsis must represent ", num_ell_idxes, " dimensions in all terms");
-          }
-          for (int64_t i = 0; i < num_ell_idxes; ++i) { // map ellipsis dimensions in operand to indices
-            current_op_idxes.push_back(first_ell_idx + i);
-            last_idx_occurrence.push_back(operand);
-          }
-          dims_in_term += num_ell_idxes;                // keep track of dimensions
-        }
-      } else {                                          // a letter (hopefully)
-        TORCH_CHECK((ell_char_count == 0) || (ell_char_count == 3), "'.' must only occur in ellipsis, operand ", operand);
-        TORCH_CHECK(('a' <= c) && (c <= 'z'), "only lowercase letters a-z allowed as indices");
-        int64_t letter_num = c-'a';                     // letter_num  = position in letter_mapping
-        if (letter_mapping[letter_num] == -1) {         // new letter, add internal index and mapping
-          letter_mapping[letter_num] = num_total_idxes;
-          num_total_idxes++;
-          last_idx_occurrence.push_back(operand);
-        } else {                                        // letter we have already seen
-          last_idx_occurrence[letter_mapping[letter_num]] = operand;
-        }
-        num_letter_occurrences[letter_num]++;
-        current_op_idxes.push_back(letter_mapping[letter_num]);
-        dims_in_term++;
-      }
+// There are roughly three parts to compute einsum:
+// 1. Parse equation to extract the labels for each input operand and output
+// 2. Unsqueeze missing dimensions from input operands and permute to align them
+// 3. Compute result by multiplying input operands and summing contraction
+//    dimensions We do the last part by reducing to bmm, there is room for
+//    optimization here.
+Tensor einsum(std::string equation, TensorList operands) {
+  TORCH_CHECK(!operands.empty(), "einsum() must provide at least one operand");
+  checkDeviceType("einsum()", operands, operands[0].device().type());
+
+  // Code for encoding ellipsis with labels (...)
+  constexpr int ELLIPSIS = '.';
+
+  // Find arrow (->) to split equation into lhs and rhs
+  std::size_t arrow_pos = equation.find("->");
+
+  // Parse lhs of equation to store ASCII value for each valid label and
+  // ELLIPSIS for every operand into op_labels
+  std::string lhs = equation.substr(0, arrow_pos);
+  std::vector<std::vector<int>> op_labels(operands.size());
+  bool found_ell = false;
+  std::size_t curr_op = 0;
+  for (std::size_t i = 0; i < lhs.length(); ++i) {
```

Comment:
Technically `i` would be `std::string::size_type`. You can get around the type mismatch by writing

```
const auto lhs_length = lhs.length()
for (auto i = decltype(lhs_length){0}; i < lhs_length; ++i) { 
```

If you really wanted to be fancy you could even use functions like `std::result_of` to avoid explicitly creating `lhs_length`, but I think it makes the code more readable.